IA-2774 add payment edition modal

quang-le opened the pull request:
Explain what problem this PR is resolving

Related JIRA tickets : IA-2774

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [ ] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Doc
In the code (jsdocs)

## Changes

- Add `useMarkPaymentAsSent` hook to handle the patch request, so the "edit" part of `useSavePaymentLot` can be used to update payments' statuses in the lot
- Add separate modal for edition with individual buttons to edit payment lot fields and without formik or yup validation
- Use value from table row to populate table and get initial state value
- update `patchRequest` of `useSavePaymentLots`: commented out actual patch request (already formatted for use with API). log body in the console for the time being
- use `patchRequest` in `PaymentLotEditionDialog` to create callbacks for "save name "and "save comment" buttons
- Name and comment are already updated using the payment lot API (no mock)
- Refactor `usePotentialPaymentsColumns` -> `usePaymentsColumns`, with a `potential` prop
- Add `useTableSelection` convenience hook
- Add `EditPaymentDialog`:
    - Update a single table row
    - API mocked at the moment 
- Add `BulkEditPaymentDialog`:
    - Update selection of rows
    - Add custom "select all" and "unselect all" buttons 
    - Add warning modal on confirm. Layout taken from org unit bulk update, but the underlying component is `ConfirmCancelDialog`

## How to test

- Create a payment lot
- Go to Payment lots
- Click the edit icon button
- Test the different fields/buttons:
    - Name and comment already call the backend so they will be updated
    - Try editing a single payment 
    - Try selecting/unselecting multiple payments, then saving them
        - There should be a warning modal with the number of payments about to be updated
        - There should be a log of the body of the POST request in the console 

## Print screen / video


https://github.com/BLSQ/iaso/assets/38907762/05aea426-cbaf-4a6c-8b30-d4f606e62740


## Notes

Has to be merged after IA-2791, as it already implements it. PR: https://github.com/BLSQ/iaso/pull/1184

quang-le committed:
IA-2769: split useSavePaymentLot

- Add useMarkPaymentAsSent hook to handle the patch request, so the "edit" part of useSavePaymentLot can be used to update payments' statuses in the lot

quang-le committed:
Merge branch 'IA-2396-payments-lots' into IA-2769_edit_payment_lots

quang-le committed:
IA_2774: add edition modal

- Add separate modal for edition with individual buttons to edit payment lot fields and without formik or yup validation
- Use value from table row to populate table and get get initial state value
- Mutation hook to be added

quang-le committed:
Merge branch 'main' into IA-2769_edit_payment_lots

quang-le committed:
Merge branch 'main' into IA-2774_add_payment_edition_modal

quang-le committed:
IA-2774: enable PATCHing name and comments

- update patchRequest of useSavePaymentLots
- use it in PaymentLotEditionDialog to create callbacks for save name and save cxomment buttons

quang-le committed:
IA-2774: add useTableSlection hook

quang-le committed:
IA-2774: add translations

quang-le committed:
IA-2774: refactor usePotentialPaymentsColumns

- rename hook, pass 'potential' param

quang-le committed:
IA-2774 enable selection in PaymentLotEditiondialog table

- edition of rows not yet implemented
- selection from single row in table
- select all and unselect all buttons enabled

quang-le committed:
IA-2774: add comment to payment lot payload

quang-le committed:
IA-2774: use generic EditIconButton for modal

quang-le committed:
IA-2791: remove duplicate EditIconButtons

- Impacted:
    - LegendBuilder
    - UserRole modal
    - User modal
    - Beneficiary type workflow changes
    - Beneficiary type workflow follow up

quang-le committed:
IA-2774: rearrange folder structure

quang-le committed:
IA-2774: update translations

quang-le committed:
IA-2774: add EditPayment dialog

- Dialog for editing a single payment
- APi is mocked for now
- Add some typing and a hook to get the statuses list
- Add translations

quang-le committed:
IA-2774: add translations

quang-le committed:
IA-2774: add BulkEditPaymentDialog

- Add BulkEditWarning component. Uses ConfirmCancelDialog, but layout based on  OrgUnitsMultiActionsDialog
- Add useBulkSavePaymentStatus. APi call is faked, but the hook is ready to send a POST request once the backend is ready
- Add StatusSelect component to avoid duplicating the select input
- Add count prop to useTableSelection and jsdoc comments

quang-le committed:
IA-2774: cleanup code

quang-le committed:
IA-2774: reorganize files

quang-le committed:
Merge branch 'IA-2791_remove_dupe_EditIconButton' into IA-2774_add_payment_edition_modal

quang-le committed:
IA-2774: pass status to bulk update (front-end)

quang-le committed:
Merge branch 'main' into IA-2774_add_payment_edition_modal

beygorghor commented on hat/assets/js/apps/Iaso/domains/payments/components/BulkEditPayment/BulkEditPaymentsDialog.tsx:
```diff
@@ -0,0 +1,82 @@
+            setIsWarningOpen(false);
+            resetSelection();
+            closeDialog();
+        });
```
Why not using onSuccess here ?

beygorghor replied:
Maybe this is because we have only the front-end part on this PR

quang-le replied:
It seemed a bit more readable to me to have keep this logic in the component, since it's all related to the components own state

beygorghor commented on hat/assets/js/apps/Iaso/domains/payments/components/BulkEditPayment/BulkEditWarning.tsx:
```diff
@@ -0,0 +1,113 @@
+        //     confirm={onConfirm}
+        //     btnDisabled={disabled}
+        //     btnVariant="text"
+        // />
```
🗑️

quang-le replied:
Yeah, I wasn't sure about that at first

beygorghor commented on hat/assets/js/apps/Iaso/domains/payments/components/EditPaymentLot/EditPaymentLotDialog.tsx:
```diff
@@ -0,0 +1,209 @@
+    );
+};
+
+type Props = { isOpen: boolean; closeDialog: () => void; paymentLot: any };
```
We have a type `PaymentLot` no ?

beygorghor commented on hat/assets/js/apps/Iaso/domains/payments/components/EditPaymentLot/EditPaymentLotDialog.tsx:
```diff
@@ -0,0 +1,209 @@
+                                </Box>
+                            </Grid>
+                            <Grid item>
+                                {/* <Box> */}
```
🗑️

beygorghor commented on hat/assets/js/apps/Iaso/domains/payments/components/EditPaymentLot/EditPaymentLotDialog.tsx:
```diff
@@ -0,0 +1,209 @@
+                                >
+                                    Unselect All
+                                </Button>
+                                {/* </Box> */}
```
🗑️

beygorghor commented on hat/assets/js/apps/Iaso/domains/payments/hooks/config/usePaymentColumns.tsx:
```diff
@@ -59,6 +64,40 @@ export const usePotentialPaymentColumns = (): Column[] => {
+                                )}
+                            </span>
+                        );
+                    },
```
only a string is enough i think

beygorghor commented on hat/assets/js/apps/Iaso/utils/table.ts:
❤️

beygorghor commented on hat/assets/js/apps/Iaso/domains/payments/hooks/requests/useSavePaymentStatus.ts:
Maybe just specify on the description this is just the frontend part

beygorghor reviewed (changes requested):
Nice job, a few remarks:

- i have a ghost dialog while saving name or comment:

https://github.com/BLSQ/iaso/assets/12494624/e5c74515-bf77-4de9-842c-2781db7242a0


- we are missing the payment status paid
- As discussed you can use a paper to display payment lot infos, remove the divider and use only one save button

quang-le committed:
IA-2774: remove BulkEditWarning

quang-le committed:
IA-2774 code review

- add "PAID" to dropdown option for payment edition
- clean up comments
- remove useless span tag in table cell

quang-le committed:
IA-2774: Add PaymentLotInfos

- replace inputs with indivudual save buttons and text lines with component using Paper and a single save button
- add style and translations

quang-le commented:
@beygorghor  Update:

![Screenshot 2024-03-15 at 18 23 46](https://github.com/BLSQ/iaso/assets/38907762/98fa6fe5-9ce9-4733-9387-2b2611a4e6af)
![Screenshot 2024-03-15 at 18 23 52](https://github.com/BLSQ/iaso/assets/38907762/8688a41f-2126-4102-8619-3186dab4acd3)
